guides/inventing-service.md

```diff
@@ -0,0 +1,919 @@
+    fn call(&mut self, request: R) -> Self::Future {
+        // 1. Check a counter for the number of requests currently being
+        //    processed.
+        // 2. If there is capacity left send the request to `T`
+        //    and increment the counter.
+        // 3. If not somehow wait until capacity becomes available.
+        // 4. When the response has been produced, decrement the counter.
+    }
+}
+```
+
+If there is no capacity left we have to wait and somehow get notified when
+capacity becomes available. Additionally we have to keep the request in memory
+while we're waiting (also called buffering). It would be more robust to only
+allocate space for the request when we are sure the service has capacity to
+handle it. Otherwise we risk using a lot of memory buffering requests while we
+wait for our service to become ready.
+
+In a way it would be nice if `Service` had a method like this:
+
+```rust
+trait Service<R> {
+    async fn ready(&mut self);
+}
+```
+
+`ready` would be an async function that completes when the service has capacity
+enough to receive one new request. We would then require users to first call
+`service.ready().await` before doing `service.call(request).await`.
+
+Separating "calling the service" from "reserving capacity" also unlocks new use
+cases such as being able to maintain a set of "ready services" that we keep up
+to date in the background such that when a request arrives we already have a
+ready service to send it to and don't have to first wait for it to become ready.
+
+With this setup `ConcurrencyLimit` could track capacity inside `ready` and not
+allow users to call `call` until there is sufficient capacity.
+
+`Service`s that don't care about capacity can just return immediately from
+`ready` or delegate to some inner `Service` they're wrapping.
+
+However we still cannot define async functions inside traits. We could add
+another associated type to `Service` called `ReadyFuture` but having to return a
+`Future` will give us the same lifetime issues we've run into before. It would
+be nice if there was someway around that.
+
+Instead we can take some inspiration from the `Future` trait and define a method
+called `poll_ready`:
+
+```rust
+use std::task::{Context, Poll};
+
+trait Service<R> {
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<()>;
+}
+```
+
+This means if the service is at capacity `poll_ready` will return
+`Poll::Pending` and somehow notify the caller, using the waker from the
+`Context`, when capacity becomes available. At that point `poll_ready` can be
+called again and if it returns `Poll::Ready(())` then capacity would be reserved
+and `call` can be called.
+
+Note that there is technically nothing that prevents users from calling `call`
+without first making sure the service is ready. Doing so however is considered a
+violation of the `Service` contract and implementations are allowed to `panic`
+if `call` is called on a service that isn't ready!
```

Comment:
Should we mention here that getting `Ready` means a slot is "reserved" for us, and that we should then use it or we're holding up capacity in the system? Plus maybe a footnote with a link to the `disarm` stuff?